cmd/kubeadm/app/cmd/phases/controlplane.go

```diff
@@ -17,186 +17,111 @@ limitations under the License.
-		cmd.Flags().StringVar(&cfg.KubernetesVersion, "kubernetes-version", cfg.KubernetesVersion, `Choose a specific Kubernetes version for the control plane`)
-
-		if properties.use == "all" || properties.use == "apiserver" {
-			cmd.Flags().StringVar(&cfg.APIEndpoint.AdvertiseAddress, "apiserver-advertise-address", cfg.APIEndpoint.AdvertiseAddress, "The IP address of the API server is accessible on")
-			cmd.Flags().Int32Var(&cfg.APIEndpoint.BindPort, "apiserver-bind-port", cfg.APIEndpoint.BindPort, "The port the API server is accessible on")
-			cmd.Flags().StringVar(&cfg.Networking.ServiceSubnet, "service-cidr", cfg.Networking.ServiceSubnet, "The range of IP address used for service VIPs")
-			cmd.Flags().StringVar(&featureGatesString, "feature-gates", featureGatesString, "A set of key=value pairs that describe feature gates for various features. "+
-				"Options are:\n"+strings.Join(features.KnownFeatures(&features.InitFeatureGates), "\n"))
-			cmd.Flags().Var(utilflag.NewMapStringString(&cfg.APIServerExtraArgs), "apiserver-extra-args", "A set of extra flags to pass to the API Server or override default ones in form of <flagname>=<value>")
-		}
-
-		if properties.use == "all" || properties.use == "controller-manager" {
-			cmd.Flags().StringVar(&cfg.Networking.PodSubnet, "pod-network-cidr", cfg.Networking.PodSubnet, "The range of IP addresses used for the Pod network")
-			cmd.Flags().Var(utilflag.NewMapStringString(&cfg.ControllerManagerExtraArgs), "controller-manager-extra-args", "A set of extra flags to pass to the Controller Manager or override default ones in form of <flagname>=<value>")
-		}
-
-		if properties.use == "all" || properties.use == "scheduler" {
-			cmd.Flags().Var(utilflag.NewMapStringString(&cfg.SchedulerExtraArgs), "scheduler-extra-args", "A set of extra flags to pass to the Scheduler or override default ones in form of <flagname>=<value>")
-		}
+type controlPlaneData interface {
+	Cfg() *kubeadmapi.InitConfiguration
+	KubeConfigDir() string
+	ManifestDir() string
+}
 
-		cmd.Flags().StringVar(&cfgPath, "config", cfgPath, "Path to kubeadm config file. WARNING: Usage of a configuration file is experimental")
+func getPhaseDescription(component string) string {
+	return fmt.Sprintf("Generates the %s static Pod manifest", component)
+}
 
-		subCmds = append(subCmds, cmd)
+// NewControlPlanePhase creates a kubeadm workflow phase that implements bootstrapping the control plane.
+func NewControlPlanePhase() workflow.Phase {
+	phase := workflow.Phase{
+		Name:    "control-plane",
+		Short:   "Generates all static Pod manifest files necessary to establish the control plane",
+		Example: controlPlaneExample,
+		Phases: []workflow.Phase{
+			newControlPlaneSubPhase(kubeadmconstants.KubeAPIServer),
+			newControlPlaneSubPhase(kubeadmconstants.KubeControllerManager),
+			newControlPlaneSubPhase(kubeadmconstants.KubeScheduler),
+		},
```

Comment:
in certs and kubeconfig I have added a run method to the top level phase that prints the folder where stuff are going to be saved. What do you think about doing the same here?

Reply:
SGTM